src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -773,38 +774,48 @@ public override BoundNode VisitArrayAccess(BoundArrayAccess node)
-                return base.VisitArrayAccess(node)!;
+                resultExpr = base.VisitArrayAccess(node)!;
             }
-
-            var indexType = VisitType(node.Indices[0].Type);
-            var F = _factory;
-
-            BoundNode resultExpr;
-            if (TypeSymbol.Equals(
-                indexType,
-                _compilation.GetWellKnownType(WellKnownType.System_Range),
-                TypeCompareKind.ConsiderEverything))
+            else
             {
-                // array[Range] is compiled to:
-                // System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray(array, Range)
-
-                Debug.Assert(node.Expression.Type is { TypeKind: TypeKind.Array });
-                var elementType = ((ArrayTypeSymbol)node.Expression.Type).ElementTypeWithAnnotations;
-
-                resultExpr = F.Call(
-                    receiver: null,
-                    F.WellKnownMethod(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__GetSubArray_T)
-                        .Construct(ImmutableArray.Create(elementType)),
-                    ImmutableArray.Create(
-                        VisitExpression(node.Expression),
-                        VisitExpression(node.Indices[0])));
+                var indexType = VisitType(node.Indices[0].Type);
+                var F = _factory;
+
+                if (TypeSymbol.Equals(
+                    indexType,
+                    _compilation.GetWellKnownType(WellKnownType.System_Range),
+                    TypeCompareKind.ConsiderEverything))
+                {
+                    // array[Range] is compiled to:
+                    // System.Runtime.CompilerServices.RuntimeHelpers.GetSubArray(array, Range)
+
+                    Debug.Assert(node.Expression.Type is { TypeKind: TypeKind.Array });
+                    var elementType = ((ArrayTypeSymbol)node.Expression.Type).ElementTypeWithAnnotations;
+
+                    resultExpr = F.Call(
+                        receiver: null,
+                        F.WellKnownMethod(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__GetSubArray_T)
+                            .Construct(ImmutableArray.Create(elementType)),
+                        ImmutableArray.Create(
+                            VisitExpression(node.Expression),
+                            VisitExpression(node.Indices[0])));
+                }
+                else
+                {
+                    resultExpr = base.VisitArrayAccess(node)!;
+                }
             }
-            else
+
+            if (_inCompoundAssignmentReceiver &&
+                node.Expression.Type is ArrayTypeSymbol { ElementType: { TypeKind: TypeKind.TypeParameter, IsValueType: false } } &&
+                resultExpr is BoundArrayAccess arr)
             {
-                resultExpr = base.VisitArrayAccess(node)!;
+                resultExpr = arr.Update(arr.Expression, arr.Indices, inCompoundAssignmentReceiver: true, arr.Type);
```

Comment:
>inCompoundAssignmentReceiver: true [](http://example.com/codeflow?start=69&length=34)

I think that, in a general case, this assumption is incorrect. The expression that we are rewriting while ```_inCompoundAssignmentReceiver``` is set to true could be arbitrarily complex, with any number of array access expressions used in it in any number of positions that are not receiver positions. More over, the array access can be a target of an explicit `ref` taking operation, like right hand side of a `ref` assignment, or a `ref` argument, etc.

Reply:
To clarify. Strictly speaking, yes, the array access is in a compound assignment receiver. However, the value of ```inCompoundAssignmentReceiver``` is used by emit layer as an indicator that using `readonly.` prefix is appropriate. Which is incorrect in a general case, even if we put the IL verification issue aside.